compiler/parser/RdrHsSyn.hs

```diff
@@ -2454,6 +2457,8 @@ mkTemplateClassInstanceSigs templateName tyVars mbKeyType =
     , ("create",    templateType `mkFunTy` mkUpdate (mkParenTy contractId))
     , ("fetch",     contractId `mkFunTy` mkUpdate (mbParenTy templateType))
     , ("archive",   contractId `mkFunTy` mkUpdate unitType)
+    , ("toAnyTemplate", templateType `mkFunTy` anyTemplateType)
+    , ("fromAnyTemplate", anyTemplateType `mkFunTy` mkParenTy (mkQualType "Optional" `mkAppTy` templateType))
```

Comment:
We might need parens around `templateType` for pretty printing if there are type variables, i.e. `mbParenTy templateType`. Not a big deal, I can add it in my next PR.

Reply:
Are you sure this is required? This seems to work fine for generic templates so it isn’t clear to me in which case it would break.

Reply:
Asking a bit more broadly, does GHC ever need `mbParenTy` around ``y `mkAppTy` z`` in ``x `mkAppTy` (y `mkAppTy` z)`` It seems weird that it would try to rearrange this after parsing given that application is builtin as opposed to user-defined operators that you have to reshuffle after parsing.

Reply:
Thinking about this a bit more, there are no user-defined precedences for types so I’m not sure why GHC wouldn’t just arrange things according to precedence during parsing and not do any kind of reshuffling on types afterwards.

I’d be very curious to know why that’s not how it’s implemented (presumably some of those `mbParenType` calls have been added since things break otherwise?) and under which circumstances they are required. Maybe @shayne-fletcher-da knows the answers.

Reply:
The `mbParenTy` calls are not strictly required - nothing breaks without them. I just added them when I observed that the pretty print of the parsed code looks wrong without them, e.g. `Optional Iou t` when the grouping of the AST is implicitly `Optional (Iou t)`. The paren nodes in the AST make the grouping explicit, but this is possibly only useful to the person testing/debugging the desugarer.